File: BotsSDK/korebot/src/main/java/kore/botssdk/audiocodes/webrtcclient/Activities/AccountActivity.java
package kore.botssdk.audiocodes.webrtcclient.Activities;

import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.audiocodes.mv.webrtcsdk.sip.enums.Transport;

import kore.botssdk.R;
import kore.botssdk.application.BotApplication;
import kore.botssdk.audiocodes.oauth.OAuthManager;
import kore.botssdk.audiocodes.webrtcclient.General.ACManager;
import kore.botssdk.audiocodes.webrtcclient.General.AppUtils;
import kore.botssdk.audiocodes.webrtcclient.General.Log;
import kore.botssdk.audiocodes.webrtcclient.General.Prefs;
import kore.botssdk.audiocodes.webrtcclient.Structure.SipAccount;


public class AccountActivity extends BaseAppCompatActivity {

    private static final String TAG = "AccountActivity";

    private boolean oAuthEnable;

    private EditText secondCall, transferCall;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //this.requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.account_activity);

        initGui();
    }

    private void initGui() {
        SipAccount sipAccount = Prefs.getSipAccount();


        EditText userNameEditText = (EditText) findViewById(R.id.account_editText_username);
        //userNameEditText.setText(sipAccount.getUsername());
        EditText passwordEditText = (EditText) findViewById(R.id.account_editText_password);
        EditText displayNameEditText = (EditText) findViewById(R.id.account_editText_displayname);
        EditText domainEditText = (EditText) findViewById(R.id.account_editText_domain);
        EditText sipAddressEditText = (EditText) findViewById(R.id.account_editText_sipaddress);
        EditText portEditText = (EditText) findViewById(R.id.account_editText_port);
        Spinner transportSpinner = (Spinner) findViewById(R.id.account_editText_transport);

        transportSpinner.setAdapter(new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, Transport.values()));
//        EditText turnServerEditText = (EditText)findViewById(R.id.login_editText_turnserver);
//        EditText stunserverEditText = (EditText)findViewById(R.id.login_editText_stunserver);

        Button loginButton = (Button) findViewById(R.id.account_button_login);

        secondCall = (EditText) findViewById(R.id.account_editText_second_call);
        transferCall = (EditText) findViewById(R.id.account_editText_transfer_call);
        secondCall.setText(Prefs.getSecondCall());
        transferCall.setText(Prefs.getTransferCall());

        TextView statusTextView = (TextView) findViewById(R.id.account_textview_status);
        String statusStr = getString(R.string.account_textview_status_prefix) + " " +
                (ACManager.getInstance().isRegisterState() ? getString(R.string.account_textview_status_connected) : getString(R.string.account_textview_status_disconnected));
        if (sipAccount != null && sipAccount.getUsername() != null)
        {
            try {
                statusStr+=", "+sipAccount.getUsername().toString();
            } catch (Exception e) {
                Log.e(TAG, "err: "+e);
            }
        }
        statusTextView.setText(statusStr);
        //boolean test = true;
        // if (test)
        //{
        userNameEditText.setText(sipAccount.getUsername());
        passwordEditText.setText(sipAccount.getPassword());
        displayNameEditText.setText(sipAccount.getDisplayName());
        domainEditText.setText(sipAccount.getDomain());
        sipAddressEditText.setText(sipAccount.getProxy());

        for (int i = 0; i < transportSpinner.getAdapter().getCount(); i++) {
            if (transportSpinner.getItemAtPosition(i).toString().toLowerCase().equals(sipAccount.getTransport().toString().toLowerCase())) {
                transportSpinner.setSelection(i);
            }
        }

        //userNameEditText.setText(sipAccount.getUsername());
        //userNameEditText.setText(sipAccount.getUsername());
        portEditText.setText(String.valueOf(sipAccount.getPort()));


        EditText oauthUrlEditText = (EditText)findViewById(R.id.login_editText_oauth_url);
        EditText oauthRealmEditText = (EditText)findViewById(R.id.login_editText_oauth_realm);
        EditText oauthClientIdEditText = (EditText)findViewById(R.id.login_editText_oauth_clientid);

        if(OAuthManager.getInstance().isEnabled()){
            oAuthEnable = true;
            LinearLayout layout = (LinearLayout) findViewById(R.id.oauth_layout);
            layout.setVisibility(View.VISIBLE);
            oauthUrlEditText.setText(OAuthManager.getInstance().getURL());
            oauthRealmEditText.setText(OAuthManager.getInstance().getRealm());
            oauthClientIdEditText.setText(OAuthManager.getInstance().getClientId());
            passwordEditText.setText("");
        }

        CheckBox autologin = (CheckBox) findViewById(R.id.autologin_checkBox);
        autologin.setChecked(Prefs.getAutoLogin());

        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String userName = userNameEditText.getText().toString().trim();
                String password = passwordEditText.getText().toString().trim();
                if (validateField(userNameEditText) && validateField(passwordEditText)) {

                    sipAccount.setUsername(userName);
                    if (!oAuthEnable) {
                        sipAccount.setPassword(password);
                    } else {
                        sipAccount.setPassword("no password");
                    }
                    //if(validateField(displayNameEditText)&&validateField(domainEditText)&&validateField(sipAddressEditText)&&validateField(turnServerEditText)&&validateField(stunserverEditText))
                    if (validateField(displayNameEditText) && validateField(domainEditText) && validateField(sipAddressEditText) && validateField(portEditText)) {
                        sipAccount.setDisplayName(displayNameEditText.getText().toString().trim());
                        sipAccount.setDomain(domainEditText.getText().toString().trim());
                        sipAccount.setProxy(sipAddressEditText.getText().toString().trim());
                        sipAccount.setTransport(AppUtils.getTransport(transportSpinner.getSelectedItem().toString().trim()));

                        int portNumber;
                        try {
                            portNumber = Integer.valueOf(portEditText.getText().toString().trim());
                        } catch (Exception e) {
                            //use default port number;
                            portNumber = Integer.valueOf(getString(R.string.sip_account_port_default));
                        }
                        sipAccount.setPort(portNumber);
                    }

                }

                Log.d(TAG, "start logout");
                Prefs.setSipAccount(sipAccount);
                handleOauth(oauthUrlEditText, oauthRealmEditText, oauthClientIdEditText,
                            userName, password, autologin.isChecked());
            }
        });
    }

    private void handleOauth(EditText oauthUrlEditText, EditText oauthRealmEditText, EditText oauthClientIdEditText, String userName, String password,
                             boolean autologin){
        boolean disconnectCall = Prefs.getDisconnectBrokenConnection();
        if(oAuthEnable){
            OAuthManager.getInstance().setURL(oauthUrlEditText.getText().toString());
            OAuthManager.getInstance().setRealm(oauthRealmEditText.getText().toString());
            OAuthManager.getInstance().setClientId(oauthClientIdEditText.getText().toString());
            OAuthManager.getInstance().authorize(userName, password, new OAuthManager.LoginCallback() {
                @Override
                public void onAuthorize(boolean success) {
                    if(success){
                        openNextScreen(autologin, disconnectCall);
                    }
                    else{
                        Toast.makeText(BotApplication.getGlobalContext(),getString(R.string.oauth_error), Toast.LENGTH_SHORT).show();
                    }
                }
            });
        }
        else{
            openNextScreen(autologin, disconnectCall);
        }
    }

    private boolean validateField(EditText editText) {
        if (editText != null && editText.getText() != null && !editText.getText().toString().trim().equals("")) {
            return true;
        }
        return false;
    }

    private void openNextScreen(boolean autologin, boolean disconnectCall) {
        Prefs.setSecondCall(secondCall.getText().toString());
        Prefs.setTransferCall(transferCall.getText().toString());
        Prefs.setFirstLogin(false);
        //start login and open app main screen
        ACManager.getInstance().startLogin(autologin, disconnectCall);
        startNextActivity(MainActivity.class);
        finish();
    }

}

File: BotSdk/ui/src/main/java/com/kore/ui/audiocodes/webrtcclient/activities/AccountActivity.kt
package com.kore.ui.audiocodes.webrtcclient.activities

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import com.audiocodes.mv.webrtcsdk.sip.enums.Transport
import com.kore.common.utils.LogUtils
import com.kore.ui.R
import com.kore.ui.audiocodes.oauth.OAuthManager
import com.kore.ui.audiocodes.webrtcclient.general.ACManager
import com.kore.ui.audiocodes.webrtcclient.general.AppUtils
import com.kore.ui.audiocodes.webrtcclient.general.Prefs

class AccountActivity : BaseAppCompatActivity() {
    private var oAuthEnable = false
    private var secondCall: EditText? = null
    private var transferCall: EditText? = null
    private val oAuthManager: OAuthManager = OAuthManager.getInstance()
    private val acManager: ACManager = ACManager.getInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //this.requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.account_activity)
        initGui()
    }

    private fun initGui() {
        val sipAccount = Prefs.getSipAccount(this)
        val userNameEditText = findViewById<View>(R.id.account_editText_username) as EditText
        val passwordEditText = findViewById<View>(R.id.account_editText_password) as EditText
        val displayNameEditText = findViewById<View>(R.id.account_editText_displayname) as EditText
        val domainEditText = findViewById<View>(R.id.account_editText_domain) as EditText
        val sipAddressEditText = findViewById<View>(R.id.account_editText_sipaddress) as EditText
        val portEditText = findViewById<View>(R.id.account_editText_port) as EditText
        val transportSpinner = findViewById<View>(R.id.account_editText_transport) as Spinner
        transportSpinner.adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, Transport.values())
        val loginButton = findViewById<View>(R.id.account_button_login) as Button
        secondCall = findViewById<View>(R.id.account_editText_second_call) as EditText
        transferCall = findViewById<View>(R.id.account_editText_transfer_call) as EditText
        secondCall?.setText(Prefs.getSecondCall(this))
        transferCall?.setText(Prefs.getTransferCall(this))
        val statusTextView = findViewById<View>(R.id.account_textview_status) as TextView
        var statusStr = getString(R.string.account_textview_status_prefix) + " " +
                if (acManager.isRegisterState) getString(R.string.account_textview_status_connected) else getString(R.string.account_textview_status_disconnected)
        if (sipAccount.username != null) {
            try {
                statusStr += ", " + sipAccount.username
            } catch (e: Exception) {
                LogUtils.e(TAG, "err: $e")
            }
        }
        statusTextView.text = statusStr
        userNameEditText.setText(sipAccount.username)
        passwordEditText.setText(sipAccount.password)
        displayNameEditText.setText(sipAccount.displayName)
        domainEditText.setText(sipAccount.domain)
        sipAddressEditText.setText(sipAccount.proxy)
        for (i in 0 until transportSpinner.adapter.count) {
            if (transportSpinner.getItemAtPosition(i).toString().equals(sipAccount.transport.toString(), ignoreCase = true)) {
                transportSpinner.setSelection(i)
            }
        }

        portEditText.setText(sipAccount.port.toString())
        val oauthUrlEditText = findViewById<View>(R.id.login_editText_oauth_url) as EditText
        val oauthRealmEditText = findViewById<View>(R.id.login_editText_oauth_realm) as EditText
        val oauthClientIdEditText = findViewById<View>(R.id.login_editText_oauth_clientid) as EditText
        if (oAuthManager.isEnabled == true) {
            oAuthEnable = true
            val layout = findViewById<View>(R.id.oauth_layout) as LinearLayout
            layout.visibility = View.VISIBLE
            oauthUrlEditText.setText(oAuthManager.getURL())
            oauthRealmEditText.setText(oAuthManager.getRealm())
            oauthClientIdEditText.setText(oAuthManager.getClientId())
            passwordEditText.setText("")
        }
        val autologin = findViewById<View>(R.id.autologin_checkBox) as CheckBox
        autologin.isChecked = Prefs.getAutoLogin(this)
        loginButton.setOnClickListener {
            val userName = userNameEditText.text.toString().trim { it <= ' ' }
            val password = passwordEditText.text.toString().trim { it <= ' ' }
            if (validateField(userNameEditText) && validateField(passwordEditText)) {
                sipAccount.username = userName
                if (!oAuthEnable) {
                    sipAccount.password = password
                } else {
                    sipAccount.password = "no password"
                }
                if (validateField(displayNameEditText) && validateField(domainEditText) && validateField(sipAddressEditText) && validateField(
                        portEditText
                    )
                ) {
                    sipAccount.displayName = displayNameEditText.text.toString().trim { it <= ' ' }
                    sipAccount.domain = domainEditText.text.toString().trim { it <= ' ' }
                    sipAccount.proxy = sipAddressEditText.text.toString().trim { it <= ' ' }
                    sipAccount.transport = AppUtils.getTransport(this, transportSpinner.selectedItem.toString().trim { it <= ' ' })
                    val portNumber: Int = try {
                        Integer.valueOf(portEditText.text.toString().trim { it <= ' ' })
                    } catch (e: Exception) {
                        //use default port number;
                        Integer.valueOf(getString(R.string.sip_account_port_default))
                    }
                    sipAccount.port = portNumber
                }
            }
            Log.d(TAG, "start logout")
            Prefs.setSipAccount(this, sipAccount)
            handleOauth(
                oauthUrlEditText, oauthRealmEditText, oauthClientIdEditText,
                userName, password, autologin.isChecked
            )
        }
    }

    private fun handleOauth(
        oauthUrlEditText: EditText, oauthRealmEditText: EditText, oauthClientIdEditText: EditText, userName: String?, password: String?,
        autologin: Boolean
    ) {
        val disconnectCall = Prefs.getDisconnectBrokenConnection(this)
        if (oAuthEnable) {
            oAuthManager.setURL(oauthUrlEditText.text.toString())
            oAuthManager.setRealm(oauthRealmEditText.text.toString())
            oAuthManager.setClientId(oauthClientIdEditText.text.toString())
            oAuthManager.authorize(this, userName, password, object : OAuthManager.LoginCallback {
                override fun onAuthorize(success: Boolean) {
                    if (success) {
                        openNextScreen(autologin, disconnectCall)
                    } else {
                        Toast.makeText(applicationContext, getString(R.string.oauth_error), Toast.LENGTH_SHORT).show()
                    }
                }
            })
        } else {
            openNextScreen(autologin, disconnectCall)
        }
    }

    private fun validateField(editText: EditText?): Boolean {
        return editText != null && editText.text != null && editText.text.toString().trim { it <= ' ' } != ""
    }

    internal fun openNextScreen(autologin: Boolean, disconnectCall: Boolean) {
        Prefs.setSecondCall(this, secondCall?.text.toString())
        Prefs.setTransferCall(this, transferCall?.text.toString())
        Prefs.setIsFirstLogin(this, false)
        //start login and open app main screen
        acManager.startLogin(this, autologin, disconnectCall)
        startNextActivity(MainActivity::class.java)
        finish()
    }

    companion object {
        private const val TAG = "AccountActivity"
    }
}